packages/core/src/utils/pathSchema.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * This module encapsulates format of the path to a deep reference.
+ */
+
+import { Asset, Entry } from 'contentful';
+
+export type UnresolvedFieldset = Array<[null, string, string?]>;
+export type Fieldset = Array<[Entry | Asset, string, string?]>;
+
+export const parseDataSourcePathIntoFieldset = (path: string): UnresolvedFieldset => {
+  const parsedPath = parseDeepPath(path);
+
+  if (null === parsedPath) {
+    throw new Error(`Cannot parse path '${path}' as deep path`);
+  }
+
+  return parsedPath.fields.map((field) => [null, field, '~locale']);
+};
+
+/**
+ * Parse path into components, supports  L1 references (one reference follow) atm.
+ * @param path from data source. eg. `/uuid123/fields/image/~locale/fields/file/~locale`
+ *                               eg. `/uuid123/fields/file/~locale/fields/title/~locale`
+ * @returns
+ */
+export const parseDataSourcePathWithL1DeepBindings = (
+  path: string
+): {
+  key: string;
+  field: string;
+  fieldLocaleQualifier: string | null;
+  referentField: string;
+  referentLocaleQualifier: string | null;
+} => {
+  const parsedPath = parseDeepPath(path);
+
+  if (null === parsedPath) {
+    // Does it throw on non-deep bindings?
+    // I think it should support non-deep bindings... too.. no? where is it used?
+    throw new Error(`Cannot parse path '${path}' as deep path`);
+  }
+
+  return {
+    key: parsedPath.key,
+    field: parsedPath.fields[0],
+    referentField: parsedPath.fields[1],
+    // probaly I don't need this
+    fieldLocaleQualifier: '~locale',
+    referentLocaleQualifier: '~locale',
+  };
+};
+
+/**
+ * Detects if paths is valid deep-path, like:
+ *  - /gV6yKXp61hfYrR7rEyKxY/fields/mainStory/~locale/fields/cover/~locale/fields/title/~locale
+ * or regular, like:
+ *  - /6J8eA60yXwdm5eyUh9fX6/fields/mainStory/~locale
+ * @returns
+ */
+export const isDeepPath = (deepPathCandidate: string): boolean => {
+  return parseDeepPath(deepPathCandidate) !== null;
+};
+
+type DeepPathParsed = {
+  key: string;
+  fields: string[]; // without locale (as it's not used in SDK atm)
+};
+
+const parseDeepPath = (deepPathCandidate: string): DeepPathParsed | null => {
+  // ALGORITHM:
+  // We start with deep path in form:
+  // /uuid123/fields/mainStory/~locale/fields/cover/~locale/fields/title/~locale
+  // First turn string into array of segments
+  //      ['', 'uuid123', 'fields', 'mainStory', '~locale', 'fields', 'cover', '~locale', 'fields', 'title', '~locale']
+  // Then group segments into chunks, where each non-initial chunk starts with 'fields'
+  //      [
+  //        [ "", "uuid123" ],
+  //        [ "fields", "mainStory", "~locale" ],
+  //        [ "fields", "cover", "~locale" ],
+  //        [ "fields", "title", "~locale" ]
```

Comment:
> Why does have to be in this structure?
That's the intermediate structure. 

The final result is: 

```
  return {
    key: initialChunk[1],
    fields: fieldChunks.map((fieldChunk) => fieldChunk[1]),  
  };

 // which would resolve exactly into what you've asked for
 return {
   key: 'pathUUID1', 
   fields: ['mainStory', 'cover', 'title']
}

```


>With that, we can maybe also simplify the extract function with searching the string for the keyword fields and then extract them. Maybe with this regex for example

Long story short: if you recall in the early commits in this PR I have implemented all this parsing with regex. I was very excited about regexes. However, as you and Sofia have suggested changes/improvements to those regexes I came to one sad realization: 
 - I already forgot what my own regexes meant (even though I head regex101 links) and it was hard to read my own
 - as you and Sofia added suggestions, it would take me quite a some  time to play with them in regex101 and see which edge cases it doesn't cover
 - the more we were adding to regexes the harder it was to read

Bottom line, I made a U-turn on regexes and decided to rewrite everything with simple, obvious code. Where we mostly split things and check simple intermediate structures. (no edge cases are hidden by confusing regexes). 

The "Algorithm" described is really transparent and very easy to debug or chagne. (eg. in case we want to add some strict validation or things or extend how paths are working). 

ps. btw I tried the  regex, but it allows mangled paths (i know that if we continue adding more to this regex, we can get strict validation too, but I wonder isn't it just better to keep it with the current transparent algorithm?)

![image](https://github.com/contentful/experience-builder/assets/113918657/b303fd7d-05e9-4b6d-9635-6fe3e1bed675)

---

ps. BTW I have adjusted the `isValidXX()` functions to use regexes (as you have suggested).